Always define cell datatype class for query printouts, refs 2586

TK-999 opened the pull request:
This PR is made in reference to: #2586 

This PR addresses or contains:
For certain inline queries which contain empty printouts, the generated HTML table cell does not contain proper `class` attribute. We must ensure that the `class` attribute is always defined, even if the cell is empty.

This PR includes:
- [ ] Tests (unit/integration)
- [x] CI build passed

Fixes #2586

TK-999 commented:
I'll take a look at the test failures

mwjames commented:
As general note:

While the change in itself isn't causing an issue, the modification of
the output is and is causing our integration tests (those that actual
test the integration with MW) to fail.

You may not be familiar with [0] but those tests are our defense line
against regressions especially those caused by MW releases and of
course those from ourselves if we change the behaviour.

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/tests/phpunit/Integration/JSONScript/README.md

PS: The JSON may seem counter intuitive to a hard-core developer but
it frees us from writing integration tests (and hereby I mean raw
PHPUnit) one by one.

Cheers

On 9/30/17, Máté Szabó <notifications@github.com> wrote:
> This PR is made in reference to: #2586
>
> This PR addresses or contains:
> For certain inline queries which contain empty printouts, the generated HTML
> table cell does not contain proper `class` attribute. We must ensure that
> the `class` attribute is always defined, even if the cell is empty.
>
> This PR includes:
> - [ ] Tests (unit/integration)
> - [ ] CI build passed
>
> Fixes #2586
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/SemanticMediaWiki/SemanticMediaWiki/pull/2712
>
> -- Commit Summary --
>
>   * Always define cell datatype class for query printouts, refs 2586
>
> -- File Changes --
>
>     M includes/queryprinters/TableResultPrinter.php (18)
>
> -- Patch Links --
>
> https://github.com/SemanticMediaWiki/SemanticMediaWiki/pull/2712.patch
> https://github.com/SemanticMediaWiki/SemanticMediaWiki/pull/2712.diff
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/SemanticMediaWiki/SemanticMediaWiki/pull/2712
>

TK-999 commented:
@mwjames Yeah, I've updated them and they should pass after I correct my trivial copy-and-paste fails :) I've added explicit sorting to attributes so that changing their order of insertion won't break tests in the future.

mwjames commented:
> I've added explicit sorting to attributes so that changing their order of insertion won't break tests in the future.

:+1: 

> Eliminate final test failure

Thanks for the perseverance to see this through, I know it can be daunting but its the only way we can keep some form of quality assurance.

TK-999 commented:
>Thanks for the perseverance to see this through, I know it can be daunting but its the only way we can keep some form of quality assurance.

Hah, no problem, it's actually a really nice declarative approach to integration tests. And congrats on the 87% coverage 😉
